File: it-agents/skills/code-architect/references/code-examples/cpp/Facade.cpp
// Facade — a simplified interface over a complex subsystem.
#include <iostream>
#include <string>

// Subsystem components, each with its own concerns.
class Loader {
public:
    std::string load(const std::string& id) const { return "data<" + id + ">"; }
};

class Validator {
public:
    bool check(const std::string& data) const { return !data.empty(); }
};

class Renderer {
public:
    std::string render(const std::string& data) const { return "[ " + data + " ]"; }
};

// Facade orchestrates the subsystem behind one method.
class ViewFacade {
    Loader loader;
    Validator validator;
    Renderer renderer;
public:
    std::string open(const std::string& id) const {
        auto data = loader.load(id);
        if (!validator.check(data)) return "<invalid>";
        return renderer.render(data);
    }
};

int main() {
    ViewFacade facade;
    std::cout << facade.open("42") << "\n";
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/ruby/facade.rb
# frozen_string_literal: true

# Facade — a simplified interface over a complex subsystem.

# Subsystem parts, each with its own narrow responsibility.
class Compressor
  def pack(data) = "packed(#{data})"
end

class Encryptor
  def encrypt(blob) = "enc(#{blob})"
end

class Uploader
  def send(payload, to:) = "#{payload} -> #{to}"
end

# Facade hides the orchestration behind one easy method.
class BackupFacade
  def initialize
    @compressor = Compressor.new
    @encryptor = Encryptor.new
    @uploader = Uploader.new
  end

  def backup(data, destination:)
    blob = @compressor.pack(data)
    secured = @encryptor.encrypt(blob)
    @uploader.send(secured, to: destination)
  end
end

if __FILE__ == $PROGRAM_NAME
  puts BackupFacade.new.backup("report.csv", destination: "s3://bucket")
end
